Multithreaded callback

sachinpc opened the pull request:
This project illustrates calling a python callback from a c-thread (posix) using cython as glue.
Since the Python interpreter is not aware of the c-threads , we need to create a Python thread state for each C-thread.
We achieve this in Cython by creating a Python thread-pool and map the callbacks coming in from the C-threads to the threads in the pool.
This code is built on top pf Stefan Behnel's Cython/Demos/callback in github. Stefan's example cheesefinder implements a callback from c , however this does not work
as-is when the callback is called from a posix thread for the reasons cited above .

mic-e commented:
I'm slightly irritated right now; I thought that declaring your `cdef` callback `with gil` would be all that's needed to safely call it from a C thread. Am I wrong?

sachinpc commented:
I too thought the same, up until I got a core dump with an error thread_state not found. 
I did some checking to find that Python actually needs us to represent that c thread as a python thread, and the Python thread needs to be swapped-in before the callback gets called. I implemented this part using Python C API , and confirmed it works. Later refactored it with  cython/Python code using the multiprocessing.threads .

mic-e commented:
Shouldn't doing that be something that Cython should do for us? e.g. I don't see why the issue couldn't occur in line 34 of the pyx file.

sachinpc commented:
Actually line 12 in .pyx seems to be helping cython in fixing the issue. The thread pool needs to be created before the c threads are created. I will attempt to provide a more thorough analysis on the generated c code . Indeed I would have hoped cython would have done that for us, hopefully in cython 2.3 😃

sachinpc commented:
Mike, I added more logs to print the posix-thread_id , python-thread_id . I removed that part of the code used to map the callback to the thread-pool.  Cython seems to be creating threads . But it does that only if a thread pool is created . A pool of 1 thread is enough  ( since cython doesn't map the function to the thread, but instead creates a new thread per posix thread ). I have a strong feeling , there could be a more efficient way of jumpstarting cython thread-allocation instead of manually creating a thread-pool of 1 thread.

scoder commented:
In order to enable C threads to call "with gil" functions, the GIL needs to be created ahead of time. It is not created by default and cannot be created from a non-main thread.

This can be done by calling `PyEval_InitThreads()` from the main thread. Cython will automatically do this if certain constructs are used in the module (e.g. thread-parallel loops), but avoids doing so in other cases, as creating the GIL comes with a permanent runtime cost in the single-threaded case.

I don't remember much of the discussions that lead to the current status, so maybe there are other cases where Cython should call the above on import. Having "with gil" functions at all may or may not suffice as an indicator. In any case, you can investigate the current workings by searching for "InitThreads" in the code base.

mic-e commented:
So, to clarify: calling `with gil` function callbacks from a non-Python thread will be undefined behavior, up until the point where I initialize the first pure-Python thread?

scoder commented:
No. That would be rather surprising semantics and worth improving on.

Let's see if @robertwb recalls some the reasoning behind the current behaviour. Might just be a bug after all. I find it surprising that `ForceInitThreads` uses `#define __PYX_FORCE_INIT_THREADS 0`, for example. Might have been intentional at some point, but the unexpected naming may have lead to inconsiderate use in other places.

sachinpc commented:
An explicit 'with thread'  seems most cythonic and beautiful in my opinion.

sachinpc commented:
In addition , currently there is no implicit regularity in Cython to enable sub-interpreter support . I dont know if this is by choice (with no additional advantage). I would have thought this is the Python version of lwp

scoder commented:
This is the wrong place to discuss this, so please refer to the cython-users mailing list if you have questions. Essentially, subinterpreters are not (really) supported by CPython, nor Cython (at all). There is currently a drive to enable this in CPython 3.6, but that's a while away and a lot of work needs to be done in that direction first. A part of the work has been done in PEP 489, as implemented in CPython 3.5. It's possible that Cython will support that in the near future. If you want to help, please ask on the cython-devel (core developers) mailing list.

scoder commented:
Regarding "with thread", I agree that a compiler directive would help (e.g. `external_threading=True`), but it would still be nice if Cython could at least get it right in the obvious cases.

sachinpc commented:
Thanks a lot for the pointers scoder, I am new to both Cython and Opensource contributions. I will definitely subscribe to the user and developer list . Would love to contribute more!
